### ISSUE TITLE
WebAssembly + GGML_MEM_ALIGN

### ISSUE THREAD

**Issue by saraalrawi** [ggerganov/ggml#653]
I built a simple autoencoder model using ggml. 

In the forward function and after running the matrix multiplication, and activation functions, I run: 

ggml_build_forward_expand(gf, out_layer_1);

Forward function: 

struct ggml_tensor *logits = forward(&model, ctx0 ,&gf ,real_data); 

where ctx0 is the context initialised with the following parameters: 

`		struct ggml_init_params params = {
			//.mem_size   = 1024ll*1024ll*1024ll, too big! cant grow memory error 
			.mem_size   = 16 * 16 * 1024 * 1024,
			.mem_buffer = NULL,
			.no_alloc   = false,
		};
		
		
		// get context0 
	 	struct ggml_context *ctx0 = ggml_init(params); 
`

and real_data is a tensor of size 200x400 filled with float numbers. 

I can train the model and run inference as well by compiling and running the compiled C++ code. 

Later, I compiled the code into WebAssembly with MAXIMUM_MEMORY=4GB  

My issue now: I receive the following error, which is an assertion gets triggered when a new object is created (probably) 

/llama.cpp/ggml.c:4719: ((uintptr_t) (ctx->mem_buffer))%GGML_MEM_ALIGN == 0

Once the program reaches: 

`
ggml_build_forward_expand(gf, out_layer_1);

`

Can anyone help? is it because of the memory size? should I allocate memory differently?

**Comment by FSSRepo** [comment]
could help you. First of all, the way you want to allocate memory to the context is wrong. ctx0 is for building the computation graph and only requires memory for tensor metadata. The tensor data should be stored in a buffer using ggml-alloc, you can see a simple example here: [simple-ggml](https://github.com/ggerganov/ggml/discussions/563).

**Comment by saraalrawi** [comment]
@FSSRepo  Thank you very much for your reply! 

The example you provided works as WebAssmbly (with little adaptation). 
I will try to apply the concept for training a neural network now. :-) 

Let's keep the issue open till I am done so I can provide an example to be used by others :-) 

Happy coding with ggml
